### PR TITLE
Make CostTracker aware of inflight transactions

### LINKED ISSUE #366: Cost Tracker - stats reporting bug
#### Problem
- cost-tracker stats are reported when the bank is frozen
- there may still be in-flight transactions which have not have costs removed/updated from tracker

#### Proposed Solution
- Find a way to delay reporting until in-flight costs have been updated or removed

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
When a leader is packing a `Bank`, transactions are added to the cost-tracker and then later updated or removed, depending on whether the transactions were committed or not. However, it is also possible for a `Bank` to get frozen while these transactions are "in-flight".

After the bank is frozen, the bank is sent over to `CostUpdateService` almost immediately. The result is that we've observed a `Bank` having its' cost details reported to metrics BEFORE the updates / removals were made to the `CostTracker` for these in-flight transactions. This causes a leader to submit metrics with an over-reported cost in comparison to what we calculate from replaying the block.

Taking over work done by @apfitzge  in https://github.com/anza-xyz/agave/pull/398 

#### Summary of Changes
- Add in-flight transaction count to CostTracker
- Increment / decrement the in-flight transaction count in BankingStage
- Make CostUpdateService wait (with a timeout) for in-flight transaction count to settle to zero

Fixes #366

### CODE REVIEW

**Comment by steviez** [comment]
Some alternatives that were considered:
- Andrew pursued an alternative option in https://github.com/anza-xyz/agave/pull/373 that would put a lock in `Bank` and have the updating / reporting thread synchronize through that
- Another possible alternative is holding the freeze lock longer. Currently, it is dropped here inside `Consumer::execute_and_commit_transactions_locked()`:
https://github.com/anza-xyz/agave/blob/01314a4fc339e94f97468c0682c53ccc613255a6/core/src/banking_stage/consumer.rs#L701
If we returned the held freeze lock to the caller (`Consumer::process_and_record_transactions_with_pre_results()`), we could process the cost updates/removals with the lock held (ie release at the end of this block):
https://github.com/anza-xyz/agave/blob/01314a4fc339e94f97468c0682c53ccc613255a6/core/src/banking_stage/consumer.rs#L497-L531
It doesn't look like a lot of stuff to hold the lock for, but in any case, does seem a bit riskier than the approach this PR does

**Comment by apfitzge** [comment]
I don't think holding the freeze lock for longer will work, we'd also have to acquire the lock earlier. Right now we only hold freeze lock during (record, commit), but not execution.

I think that approach is riskier and not sure we should be backporting that.
A long execution, possibly due to bug?, would then prevent other threads from moving on to the next bank.

**Review by apfitzge — DISMISSED** [review]
lgtm - let's get approval from second set of eyes before merging.

**Comment by steviez** [comment]
> I don't think holding the freeze lock for longer will work, we'd also have to acquire the lock earlier. Right now we only hold freeze lock during (record, commit), but not execution.

Ah right, thank you. So yeah, VERY significantly extending the duration of holding that lock

> I think that approach is riskier and not sure we should be backporting that. A long execution, possibly due to bug?, would then prevent other threads from moving on to the next bank.

Yep, 100% agreed on that approach being much riskier

**Comment by steviez** [comment]
> lgtm - let's get approval from second set of eyes before merging.

@t-nelson - I think you've followed along with the issue this PR aims to address so adding you as a reviewer with Tao being OOO this week

File: cost-model/src/cost_tracker.rs
Comment (bw-solana) [inline cost-model/src/cost_tracker.rs:120]: wen `saturating_sub_assign` macro?

File: core/src/cost_update_service.rs
Comment (bw-solana) [inline core/src/cost_update_service.rs]: Do we know how often in flight transaction count is above zero when we come in here? Like is this happening most of the time?

If in flight is 0 most of the time, I'm wondering if we can restructure things to only acquire the read lock once for that typical case. Something like:
```rust
loop {
    {
        let cost_tracker = bank.read_cost_tracker().unwrap();
        let in_flight_transaction_count = cost_tracker
            .in_flight_transaction_count();
        if in_flight_transaction_count == 0 || loop_count >= MAX_LOOP_COUNT {
            let slot = bank.slot();
            trace!(
                "inflight transaction count for slot {slot} at {in_flight_transaction_count} \
                after {loop_count} iterations"
            );
            cost_tracker.report_stats(bank.slot());
            break;
        }
        loop_count += 1;
    }

    std::thread::sleep(LOOP_LIMITER);
}
```

File: core/src/banking_stage/qos_service.rs
Comment (bw-solana) [inline core/src/banking_stage/qos_service.rs:178]: Want to make sure I understand why the increment is here. I believe it is because:

1. By coming in here, it indicates we have a mix of committed and uncommitted transactions (as opposed to remove_transaction_costs, where everything passed in was not committed)
2. We only add transaction to in flight if it makes it past all the error checks. Thus, we only want to include `tx_cost` that are `Ok` and not `Err`.
3. We don't care about Committed vs. NotCommitted because either way they were in flight and are no longer in flight.

File: core/src/banking_stage/qos_service.rs
Comment (bw-solana) [inline core/src/banking_stage/qos_service.rs:171]: maybe name this `num_tx_no_longer_in_flight`?

File: core/src/banking_stage/qos_service.rs
Comment (bw-solana) [inline core/src/banking_stage/qos_service.rs:213]: Could rename this one as well

**Review by bw-solana — COMMENTED** [review]
Overall LGTM. I think we could optimize the read locking on the cost reporting

File: cost-model/src/cost_tracker.rs
Comment (steviez) [reply cost-model/src/cost_tracker.rs:120]: Funny enough, the macro that `saturating_add_assign` is meant to address became stable as of Rust 1.74.0

Outside of the scope of this PR, but I'll take a stab at using the language provided mechanism over our custom macro. After all, `git blame` says Trent asked us to haha:
https://github.com/anza-xyz/agave/blob/10d06773cd1daf251aeb7a6283da6bd8d75f571a/sdk/src/lib.rs#L168-L169

File: core/src/banking_stage/qos_service.rs
Comment (steviez) [reply core/src/banking_stage/qos_service.rs:178]: 1. Exactly
2. Correct. Similar to above, we could have a mix of tx's that fit / did not fit. This is determined by this snippet:
https://github.com/anza-xyz/agave/blob/10d06773cd1daf251aeb7a6283da6bd8d75f571a/core/src/banking_stage/qos_service.rs#L106-L116
3. Almost but not exactly (code snippet further down shows the update case):
    - Transactions that are not committed result in removing the cost completely
    - Transactions that are committed result in updating the cost to use actual instead of estimated
https://github.com/anza-xyz/agave/blob/10d06773cd1daf251aeb7a6283da6bd8d75f571a/core/src/banking_stage/qos_service.rs#L192-L200

File: core/src/cost_update_service.rs
Comment (steviez) [reply core/src/cost_update_service.rs]: I don't think we have numbers for how often, but in any case, I agree with your sentiment that we can rework to avoid repeated lock grabs. I'll try to work this in.

Unrelated but as an FYI - GH allows you do syntax highlighting for code blocks. To do so, type out the following
> \```rust
> desired code block
> ...
> \```

For example,

```rust
loop {
    // Grab the lock once here
    let cost_tracker = bank.read_cost_tracker().unwrap();
```

File: core/src/banking_stage/qos_service.rs
Comment (steviez) [reply core/src/banking_stage/qos_service.rs:171]: I think I'm inclined to leave these as-is; `num_included` makes it match the variable naming here:
https://github.com/anza-xyz/agave/blob/182d27f718e1d6df347bdbe60edaba6026f1c6b5/core/src/banking_stage/qos_service.rs#L101

I think sticking with `num_included` helps reinforce that we're excluding tx's that were NOT added to the `CostTracker` initially

**Review by apfitzge — DISMISSED** [review]

**Review by bw-solana — DISMISSED** [review]
LGTM

**Comment by steviez** [comment]
https://github.com/anza-xyz/agave/pull/437/commits/c30bbbb75b6c446788e4a411cc9a7c42c455cd92 😬

**Review by apfitzge — APPROVED** [review]
lgtm - how did I misss the report remove 🤦

**Review by bw-solana — APPROVED** [review]

**Pull request merged**
